File: framework/fit/python/fit_flowable/transform_operation/map_publisher_decorator.py
# -- encoding: utf-8 --
# Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
"""
功 能：对于数据进行转换的发布者的实现。
"""
from typing import Callable, Any

from fit_flowable.core.publisher import Publisher
from fit_flowable.core.subscriber import Subscriber
from fit_flowable.operation.abstract_operation import AbstractOperation
from fit_flowable.pubilsher.abstract_publisher import AbstractPublisher


class MapPublisherDecorator(AbstractPublisher):
    """
    对于数据进行转换的发布者的实现。
    """

    class MapOperation(AbstractOperation):
        def __init__(self, mapper: Callable[[Any], Any], subscriber: Subscriber):
            super().__init__(subscriber)
            self._mapper: Callable[[Any], Any] = mapper
            self._subscriber: Subscriber = subscriber

        def do_consume(self, data) -> None:
            try:
                result = self._mapper(data)
                self.get_next_subscriber().consume(result)
            except Exception as cause:
                self.get_pre_subscription().cancel()
                self.get_next_subscriber().fail(cause)

    def __init__(self, decorated: Publisher, mapper: Callable[[Any], Any]):
        """
        构造将元素进行转换的发布者。

        :param decorated: 待进行元素转换的发布者。
        :param mapper: 指定的元素转换方式。
        """
        self._decorated = decorated
        self._mapper = mapper

    def do_subscribe(self, subscriber: Subscriber) -> None:
        self._decorated.subscribe(MapPublisherDecorator.MapOperation(self._mapper, subscriber))

File: framework/fit/java/fit-reactor/src/main/java/modelengine/fitframework/flowable/publisher/MapPublisherDecorator.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) 2024 Huawei Technologies Co., Ltd. All rights reserved.
 *  This file is a part of the ModelEngine Project.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

package modelengine.fitframework.flowable.publisher;

import static modelengine.fitframework.inspection.Validation.notNull;

import modelengine.fitframework.flowable.Publisher;
import modelengine.fitframework.flowable.Subscriber;
import modelengine.fitframework.flowable.Subscription;
import modelengine.fitframework.flowable.operation.AbstractOperation;

import java.util.function.Function;

/**
 * 表示 {@link Publisher} 的携带数据类型转换的实现。
 *
 * @param <T> 表示发布者中数据转换前的类型的 {@link T}。
 * @param <R> 表示发布者中数据最终类型的 {@link R}。
 * @author 季聿阶
 * @since 2024-02-10
 */
public class MapPublisherDecorator<T, R> implements Publisher<R> {
    private final Publisher<T> decorated;
    private final Function<T, R> mapper;

    public MapPublisherDecorator(Publisher<T> decorated, Function<T, R> mapper) {
        this.decorated = notNull(decorated, "The decorated map publisher cannot be null.");
        this.mapper = notNull(mapper, "The mapper cannot be null.");
    }

    @Override
    public void subscribe(Subscriber<R> subscriber) {
        this.decorated.subscribe(new MapOperation<>(this.mapper, subscriber));
    }

    private static class MapOperation<T, R> extends AbstractOperation<T, R> {
        private final Function<T, R> mapper;

        /**
         * 使用数据映射方法和下游的订阅者来初始化 {@link MapOperation}。
         *
         * @param mapper 表示数据映射方法的 {@link Function}{@code <}{@link T}{@code , }{@link R}{@code >}。
         * @param subscriber 表示下游的订阅者的 {@link Subscriber}{@code <}{@link T}{@code >}。
         */
        MapOperation(Function<T, R> mapper, Subscriber<R> subscriber) {
            super(subscriber);
            this.mapper = mapper;
        }

        @Override
        protected void consume0(Subscription subscription, T data) {
            try {
                R result = this.mapper.apply(data);
                this.getNextSubscriber().consume(result);
            } catch (Exception cause) {
                this.getPreSubscription().cancel();
                this.getNextSubscriber().fail(cause);
            }
        }
    }
}
